Add RPC support to sql/driver.

petermattis opened the pull request:
RPC is significantly faster than HTTP:

```
name         old time/op     new time/op    delta
Select1-8    210us +- 1%     127us +- 1%    -39.54%  (p=0.008 n=5+5)
Select1S-8   232us +- 1%     135us +- 2%    -41.90%  (p=0.008 n=5+5)
```

The "S" suffix is for secure. This is on go1.5. I also tested go-tip and
the improvement is comparable, though the overall perf numbers are about
10% lower on tip across the board (i.e. the go-tip SSL improvements are
being drowned by something else).

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/2994)

<!-- Reviewable:end -->

petermattis committed:
Add RPC support to sql/driver.

RPC is significantly faster than HTTP:

name         old time/op     new time/op    delta
Select1-8    210us +- 1%     127us +- 1%    -39.54%  (p=0.008 n=5+5)
Select1S-8   232us +- 1%     135us +- 2%    -41.90%  (p=0.008 n=5+5)

The "S" suffix is for secure. This is on go1.5. I also tested go-tip and
the improvement is comparable, though the overall perf numbers are about
10% lower on tip across the board (i.e. the go-tip SSL improvements are
being drowned by something else).

petermattis commented:
I need to add some tests that use the RPC support. Or switch the default.

tschottdorf commented on sql/driver/driver.go:
```diff
@@ -24,8 +24,14 @@ import (
+// Stopper exposed to allow tests to properly drain client connections.
+//
+// TODO(pmattis): Is there a better way to expose this?
+var Stopper *stop.Stopper
```
if this is only used for tests, can you move it into the test code?

tschottdorf replied:
Also, I don't see this being populated anywhere. What am I missing?

tschottdorf replied:
nevermind, found it.

petermattis replied:
It is used by tests to cleanly shut down client goroutines. I'm not really fond of the way this is working but we have restrictions from the `database/sql` interface.

tschottdorf commented on sql/driver/link_test.go:
```diff
@@ -26,6 +26,8 @@ import (
 )
 
 func TestNoLinkForbidden(t *testing.T) {
+	t.Skipf("broken for now")
```
the comment isn't too helpful. Can you be more verbose?

petermattis replied:
Yeah, we're linking it stuff now that was previously forbidden. I need to fix the test to accept those new imports. Just forgot to do so before sending out the pull-request.

tamird commented on sql/server.go:
```diff
@@ -26,24 +26,26 @@ import (
 	"github.com/cockroachdb/cockroach/sql/driver"
 	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/gogo/protobuf/proto"
```
nit: this should go above the cockroachdb imports

tamird commented on testutils/build.go:
```diff
@@ -31,6 +31,9 @@ func TransitiveImports(importpath string, cgo bool) (map[string]struct{}, error)
 	addImports = func(root string) error {
 		pkg, err := buildContext.Import(root, buildContext.GOPATH, 0)
 		if err != nil {
+			if _, ok := err.(*build.NoGoError); ok {
```
what's this?

tamird commented on sql/driver/rpc_sender.go:
```diff
@@ -0,0 +1,110 @@
+	RegisterSender("rpcs", f)
+}
+
+const method = "Server.SQL"
```
export this so you can use it in sql/server.go

tamird commented on sql/driver/driver_test.go:
```diff
@@ -385,28 +386,38 @@ func TestConnectionSettings(t *testing.T) {
-		t.Fatal(err)
+
+	// Test that all of the network protocols work.
+	for _, scheme := range []string{"http", "https", "rpc", "rpcs"} {
```
FWIW we already have constants for these in base/context.go; I think it'd be better to export them there so this doesn't rot.

tamird commented on sql/driver/driver_test.go:
```diff
@@ -385,28 +386,38 @@ func TestConnectionSettings(t *testing.T) {
+			}
+			defer s.Stop()
+
+			driver.Stopper = stop.NewStopper()
```
nit: this duplicates driverInit, which would be good to reuse.

tamird commented on sql/bench_test.go:
```diff
@@ -46,4 +63,21 @@ func BenchmarkSelect1(b *testing.B) {
+}
+
+func BenchmarkSelect1_HTTP(b *testing.B) {
+	benchmarkSelect1(b, "http")
```
ditto re constants

tamird commented:
code LGTM but I don't think we want to go down this road in practice; perhaps we should push the gRPC branch over the finish line instead?

petermattis commented:
Do you still have the gRPC branch around? I'd be happy to try benchmarking that as well. And I'm also going to benchmark the pgwire branch.
<!-- Reviewable comment -K2DHSn0-rlS0lW9q-Rw:1079462075 -->

---

Review status: 0 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[sql/bench_test.go, line 70 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jys-r2-70)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/bench_test.go#L70)):</sup>
See my other comment.
<!-- Reviewable comment -K2DHPd2MTkU6IH9SQuh:130912033 -->

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
DRY is overrated. These constants are very unlikely to change. Even if they were exported from `base/context.go` we'd have to take additional care that a new constant is not added. I struggle to see a net benefit here.
<!-- Reviewable comment -K2DHDR3nCtTWvYnbNSg:887322733 -->

---

<sup>**[sql/driver/driver_test.go, line 404 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-404)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L404)):</sup>
DRY is overrated.
<!-- Reviewable comment -K2DGBSQWxPL6LjZpASB:1882611578 -->

---

<sup>**[sql/driver/rpc_sender.go, line 42 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyz-r2-42)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/rpc_sender.go#L42)):</sup>
Such little benefit for an additional dependency between packages. And we don't attempt this sharing elsewhere (e.g. https://github.com/cockroachdb/cockroach/blob/master/client/rpc_sender.go#L45), nor do I think we should.
<!-- Reviewable comment -K2DFxw7P9Sc_rJPyWiE:-1085662277 -->

---

<sup>**[sql/server.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jz--r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/server.go#L34)):</sup>
That's the way goimports formats this stuff. We certainly aren't strict about putting it before cockroach imports elsewhere.
<!-- Reviewable comment -K2DFKG0sjddFJve2-uL:-1172732970 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
`cockroach/c-lz4` is being linked into the client and that package returns a no buildable go files error (a.k.a. `build.NoGoError`).
<!-- Reviewable comment -K2DF7W-Ja7GAgSuchVo:-90954304 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

vivekmenezes commented on sql/driver/rpc_sender.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	if context.Insecure {
+		log.Warning("running in insecure mode, this is strongly discouraged. See --insecure and --certs.")
```
can we not print this when we are in dev mode.

tamird commented:
I sure do https://github.com/tamird/cockroach/tree/sql-grpc
<!-- Reviewable comment -K2DIkvNCIiGGMt-nb6D:-79564316 -->

---

Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
I don't think it's overrated. Just yesterday I submitted a PR that fixed benchmarks that had rotted in exactly this way.
<!-- Reviewable comment -K2DIuHtthwZEP2mYOwt:-1929222767 -->

---

<sup>**[sql/driver/rpc_sender.go, line 42 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyz-r2-42)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/rpc_sender.go#L42)):</sup>
Isn't this dependency already in place?
<!-- Reviewable comment -K2DJ2MbSuTdAAyn5z5y:-1242011164 -->

---

<sup>**[sql/server.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jz--r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/server.go#L34)):</sup>
goimports is agnostic; it does this by default but if you separate them by a space it will honour your choice. I'm pretty sure we try to organize things elsewhere.
<!-- Reviewable comment -K2DJ6KPCrOuiULm1PS3:1471178632 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
That's worth a comment, I think.
<!-- Reviewable comment -K2DJFpKDf0HQqJQcX5y:478835213 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
Which PR are you referring to? The `client/client_test.go` change was something different. Wasn't it?
<!-- Reviewable comment -K2DYVBSycywkNPDS8jZ:-2067673674 -->

---

<sup>**[sql/driver/rpc_sender.go, line 42 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyz-r2-42)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/rpc_sender.go#L42)):</sup>
Yeah, the dep is already there. Feels weird for the driver to be defining the name of the endpoint for the server to use and yet it would be silly to introduce an additional package both `sql` and `sql/driver` could point at just to DRY this up.  Changed.
<!-- Reviewable comment -K2DZ3cSzbEGiQD7OKTN:2037941865 -->

---

<sup>**[sql/driver/rpc_sender.go, line 60 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyz-r3-60)** ([raw file](https://github.com/cockroachdb/cockroach/blob/fa08a6fed5a18f5b83a31ea462b4e5db4c470fdd/sql/driver/rpc_sender.go#L60)):</sup>
We don't know about dev mode here. But there is likely cleanup work to be done as fallout from this PR. I think that's ok as we're getting started on performance work.
<!-- Reviewable comment -K2DYitEpHHllj1p5gGa:-103162033 -->

---

<sup>**[sql/server.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jz--r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/server.go#L34)):</sup>
`git grep` says we're very inconsistent.
<!-- Reviewable comment -K2DYyEYbp5l6ue4EBkT:2069718522 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
Sure.
<!-- Reviewable comment -K2DZ-cGH6atJVQ00FOw:-485110604 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
The rot in #2986 was not because the benchmark was using string literals instead of constants, it was because it was using a constant instead of a value from the context.  Simply exporting and using `base.RPCScheme` and friends wouldn't help there. 
<!-- Reviewable comment -K2DhRtAPbnCgCT95cXi:-1234420903 -->

---

<sup>**[sql/driver/driver_test.go, line 404 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-404)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L404)):</sup>
I agree with Tamir here; we should do this dance in one place.  Maybe we could even unexport `driver.Stopper` and replace it with a `driver.Drain` function that would stop the stopper and replace it with a new one.
<!-- Reviewable comment -K2DhwoHM7tuCXS0TyMW:-257169378 -->

---

<sup>**[sql/driver/link_test.go, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyv-r1-29)** ([raw file](https://github.com/cockroachdb/cockroach/blob/ad23207b65f730de5b279309d22f0294e2429421/sql/driver/link_test.go#L29)):</sup>
It's forbidden for a reason: we can't let packages that define flags leak into client code.
<!-- Reviewable comment -K2E5Tbj5Emoz1xvajtU:-189190248 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
Does `cgo_flags.go` not count? Could we add some dummy code there so it doesn't confuse build introspection tools like this?
<!-- Reviewable comment -K2E4vbk-YfEUd94jLds:-2063423173 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

<sup>**[sql/driver/link_test.go, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyv-r1-29)** ([raw file](https://github.com/cockroachdb/cockroach/blob/ad23207b65f730de5b279309d22f0294e2429421/sql/driver/link_test.go#L29)):</sup>
I don't think we can use `cockroach/rpc` in the client then. Perhaps I can use `cockroach/rpc/codec` directly instead. We lose the heartbeat functionality, but does the SQL client need/want that?
<!-- Reviewable comment -K2E6jSQl7_ESCuFrSRs:-1468959445 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
Apparently it doesn't. Let me investigate more.
<!-- Reviewable comment -K2E6doePM-NapL2VRaN:762898985 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[sql/driver/link_test.go, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyv-r1-29)** ([raw file](https://github.com/cockroachdb/cockroach/blob/ad23207b65f730de5b279309d22f0294e2429421/sql/driver/link_test.go#L29)):</sup>
I think the client would like to have some sort of health checks, although it definitely doesn't need the full time-synchronization logic of our full heartbeats. Using codec directly would probably be fine; we could implement a simple ping separately.
<!-- Reviewable comment -K2E7_j4AANAO1OarxJc:-486970791 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[sql/driver/driver_test.go, line 404 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-404)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L404)):</sup>
Switching to `rpc/codec` gets rid of the stopper stuff.
<!-- Reviewable comment -K2EAZTwZOkUX5i7Vd4p:1592560667 -->

---

<sup>**[sql/driver/link_test.go, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyv-r1-29)** ([raw file](https://github.com/cockroachdb/cockroach/blob/ad23207b65f730de5b279309d22f0294e2429421/sql/driver/link_test.go#L29)):</sup>
I've switched to using codec directly which got rid of all of the additional deps. A minor bit of code was duplicated from `rpc/tls.go`. Getting rid of the extra deps seems worth the duplication which we can fix later if we feel strongly. I've highlighted the duplication with TODO comments.
<!-- Reviewable comment -K2EAd5pU0IyUOlBu2Ss:2125125482 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
The failing test was passing `cgo == false` so `cgo_flags.go` was being stripped from the list of buildable go sources. The driver now uses cgo for c-snappy and c-lz4 so I changed the test to pass `cgo == true`. 
<!-- Reviewable comment -K2EBLftbu31XUzT7DXS:-618235060 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

<sup>**[sql/driver/rpc_sender.go, line 50 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2EBvpOwE4ksbWzkVqC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/935d6ca22bdd506ba6e3f755da955bfdc4b10805/sql/driver/rpc_sender.go#L50)):</sup>
I suppose I could move this code into `rpc/codec` so that it could be shared by `rpc` and `sql/driver`. 
<!-- Reviewable comment -K2EBvpOwE4ksbWzkVqD:1169405762 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
I think we also want to avoid a (mandatory) dependency on cgo in the client. For example, windows users have trouble building things using cgo because they have to install the right version of the right compiler. Any use of cgo in the client should at least have a build tag that can be used to turn it on or off.
<!-- Reviewable comment -K2EChvYkepE3xdopgar:1052327156 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
I agree. I'll file an issue about this. I know how to fix `rpc/codec` to fall back to the Go native snappy/lz4, but don't want to clutter this PR.
<!-- Reviewable comment -K2EDsihwnPXqfbX19Og:951047878 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 10 files at r1, 8 of 10 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
On the other hand, using the already-available methods on the context in this test, would prevent that class of rot here.
<!-- Reviewable comment -K2EC_b1RhZp-JnlKo1b:324225615 -->

---

<sup>**[sql/driver/link_test.go, line 34 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2EE6WdphhaAnQuF4-n)** ([raw file](https://github.com/cockroachdb/cockroach/blob/935d6ca22bdd506ba6e3f755da955bfdc4b10805/sql/driver/link_test.go#L34)):</sup>
Just so we don't lose the discussion from elsewhere: this is an important change that we shouldn't make lightly.
<!-- Reviewable comment -K2EE6WeDOcT8xizGYId:-1091064464 -->

---

<sup>**[sql/driver/rpc_sender.go, line 50 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2EBvpOwE4ksbWzkVqC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/935d6ca22bdd506ba6e3f755da955bfdc4b10805/sql/driver/rpc_sender.go#L50)):</sup>
Yeah, I think that'd be good. Looks like the constant is already there, just not exported.
<!-- Reviewable comment -K2EDNrm-GSSIEEIkhxB:-510686107 -->

---

<sup>**[sql/server.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jz--r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/server.go#L34)):</sup>
:grapes: 
<!-- Reviewable comment -K2EDXhRyXdn-k-SIO41:1765233795 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
Ah, yes, agree with Ben on this. I've been playing with the Windows stuff on the side, and c-snappy still doesn't work properly.
<!-- Reviewable comment -K2EDqHeC5QZ8YYVINLK:-819507803 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM
<!-- Reviewable comment -K2EEVt8BBcjzhfM2ZT9:676834233 -->

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
What methods on the context could I use to replace the above slice?
<!-- Reviewable comment -K2EEa6PuQwCqX56dvDI:2133311589 -->

---

<sup>**[testutils/build.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2D4iii5nU2jYRN_jyy-r2-34)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/testutils/build.go#L34)):</sup>
I don't think Windows support should be used as a barrier to performance work at this point. Or put another way: who cares about Windows support if our performance sucks. 
<!-- Reviewable comment -K2EEvjHQtKlCqTgG3qL:-920210745 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM
<!-- Reviewable comment -K2EJtwBYk9xjtCyiHv8:676834233 -->

---

Reviewed 6 of 6 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
As discussed in person, they'd be `RPCRequestScheme` and `HTTPRequestScheme`
<!-- Reviewable comment -K2EJR7avpUI0YAhVHXi:-750297448 -->

---

<sup>**[sql/driver/link_test.go, line 34 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2EE6WdphhaAnQuF4-n)** ([raw file](https://github.com/cockroachdb/cockroach/blob/935d6ca22bdd506ba6e3f755da955bfdc4b10805/sql/driver/link_test.go#L34)):</sup>
We should at least have a comment here that we want to pass false eventually.
<!-- Reviewable comment -K2EJbJT4cRrGDs06Agw:-1791064135 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 5 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[sql/driver/link_test.go, line 34 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2EE6WdphhaAnQuF4-n)** ([raw file](https://github.com/cockroachdb/cockroach/blob/935d6ca22bdd506ba6e3f755da955bfdc4b10805/sql/driver/link_test.go#L34)):</sup>
Done.
<!-- Reviewable comment -K2I3EJb_XPpEWmobUx8:-499146009 -->

---

<sup>**[sql/driver/rpc_sender.go, line 50 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2EBvpOwE4ksbWzkVqC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/935d6ca22bdd506ba6e3f755da955bfdc4b10805/sql/driver/rpc_sender.go#L50)):</sup>
Done.
<!-- Reviewable comment -K2I3GkgJ3LEMVRRAsXn:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 6 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
Yes. I'm not actually sure this is an improvement though:

```
	for _, insecure := range []bool{false, true} {
		ctx := server.NewTestContext()
		ctx.Insecure = insecure
		for _, scheme := range []string{ctx.HTTPRequestScheme(), ctx.RPCRequestScheme()} {
```

Which protocols are being tested?
<!-- Reviewable comment -K2ISZfiAAYrh1dHSnyu:-242305647 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[sql/driver/driver_test.go, line 393 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994#-K2DDoBHI8Osm7LLG0Ap-r2-393)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31e09b8be1eef039809b2bf5cadaa43281a9ae93/sql/driver/driver_test.go#L393)):</sup>
Well I think that's a pretty trivial thing to look up, but like I said in person, I don't feel strongly any longer.
<!-- Reviewable comment -K2IUGHK72SRkLrMQwMK:639260926 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2994)
<!-- Sent from Reviewable.io -->
